src/main/java/com/google/sps/utility/GmailUtility.java

```diff
@@ -63,10 +65,21 @@ public static MessagePartHeader extractHeader(Message message, String headerName
    * @param fromHeaderValue the value of a "From" header from which a contact name / email should be
    *     extracted
    * @return A contact name if available, or an email if it is not
+   * @throws IllegalArgumentException If value could not be parsed from header
    */
   public static String parseNameInFromHeader(String fromHeaderValue) {
-    return fromHeaderValue.charAt(0) == '<'
-        ? fromHeaderValue.substring(1, fromHeaderValue.length() - 1)
-        : fromHeaderValue.split("<")[0].trim();
+    Pattern withNamePattern = Pattern.compile("(.*) <.*>");
+    Matcher withNameMatcher = withNamePattern.matcher(fromHeaderValue);
+    if (withNameMatcher.find()) {
+      return withNameMatcher.group(1);
+    }
+
+    Pattern withoutNamePattern = Pattern.compile("<(.*)>");
```

Comment:
I think you could do this with a single pattern: `(.*)?<(.*)>`
Use `groupCount` to check if there are 1 or 2 groups in the match. Don't forget to trim the name afterwards. (You could make it more complex to avoid the need to trim, but trimming is probably a good idea anyway.)

Reply:
Good suggestion - I can't do this check directly with `groupCount`, as group count simply returns the number of capturing groups in the pattern, which is always 2. However, I was able to do make this work by extracting the name group and email group, and returning the appropriate one depending on if the extracted name is empty.